File: 1079._Letter_Tile_Possibilities.py
# https://leetcode.com/problems/letter-tile-possibilities/description/?envType=daily-question&envId=2025-02-17
# Solution taken from: https://www.youtube.com/watch?v=8FrJX-P_DnE&ab_channel=NeetCodeIO
# Did I understand properly? No

class Solution:
    def numTilePossibilities(self, tiles: str) -> int:
        count = Counter(tiles)  # char -> available count
        # A, B

        # A, AB, B, BA
        def backtrack():
            res = 0

            for c in count:
                if count[c] > 0:
                    count[c] -= 1
                    res += 1
                    res += backtrack()
                    count[c] += 1
            return res

        return backtrack()

File: 1079._Letter_Tile_Possibilities.cpp
// https://leetcode.com/problems/letter-tile-possibilities/description/?envType=daily-question&envId=2025-02-17
// Solution taken from: https://www.youtube.com/watch?v=k3XBSQTnETU&ab_channel=Techdose
// Did I understand properly? No

class Solution
{
    int buildString(vector<int> &freq)
    {
        int ways = 0;
        for (int i = 0; i < 26; i++)
        {
            if (freq[i] > 0)
            {
                freq[i]--;
                ways += 1 + buildString(freq);
                freq[i]++;
            }
        }
        return ways;
    }

public:
    int numTilePossibilities(string tiles)
    {
        vector<int> freq(26, 0);
        for (char c : tiles)
            freq[c - 'A']++;
        return buildString(freq);
    }
};